backend/src/main/java/wooteco/retrospective/application/member/MemberService.java

```diff
@@ -0,0 +1,30 @@
+package wooteco.retrospective.application.member;
+
+import org.springframework.stereotype.Service;
+import wooteco.dto.member.MemberLoginRequest;
+import wooteco.dto.member.MemberLoginResponse;
+import wooteco.exception.member.MemberNotFoundException;
+import wooteco.retrospective.dao.member.MemberDao;
+import wooteco.retrospective.domain.member.Member;
+
+import java.util.Optional;
+
+@Service
+public class MemberService {
+
+    private final MemberDao memberDao;
+
+    public MemberService(MemberDao memberDao) {
+        this.memberDao = memberDao;
+    }
+
+    public MemberLoginResponse loginMember(MemberLoginRequest request) {
+        if (memberDao.exists(request.getName())) {
+            Optional<Member> member = memberDao.findByName(request.getName());
+            return MemberLoginResponse.of(member.orElseThrow(MemberNotFoundException::new));
+        }
+        Member member = memberDao.insert(request.toMember());
+        return MemberLoginResponse.of(member);
+    }
```

Comment:
```suggestion
    public MemberLoginResponse loginMember(MemberLoginRequest request) {
        return memberDao.findByName()
                .map(MemberLoginResponse::of)
                .orElseGet(
                        () -> {
                            Member member = memberDao.insert(request.toMember());
                            return MemberLoginResponse.of(member);
                        }
                );
    }
```

개인적으로 Optional을 사용하는걸 안좋아하는데요. 오용되는 코드를 작성할 가능성이 너무 높다고 느껴져요.
Optional을 반환하고 위의 로직을 수행한다면 위처럼 Optional의 Stream을 이용해서 inline으로 처리가 가능해 보입니다.
클린코드 원칙을 좀 더 지키기 위해서는 orElseGet구문을 함수로 분리하는게 더 좋아보이구요:)

Reply:
음 sugguetion같은 경우 저는 기존 코드가 읽기 좋다고 생각합니다. 
orEleseGet보다 결과물을 변환하는 행위인 map이 먼저 나오기 때문에 `멤버가 있으면 그 멤버를 돌려주고 없으면 회원가입한다` 라는 서비스 흐름의 맥이 한 번에 읽히지 않는거같아요

Reply:
갠적으로 그러면 optional을 사용하면 안됐다고 생각해요.. 옵셔널 오용의 대표적인 예 중 하나라고 생각하거든요..

저는 옵셔널 스트림의 가독성이 크게 나쁘지 않다고 보이는데...

위 로직의 경우 exist와 옵셔널이 혼용되어 사용돼서 같은 의미를 가지는 코드가 중복으로 포함되어있는 점도 문제점이라구 생각합니다!

Reply:
우선은 [이렇게](backend/src/main/java/wooteco/retrospective/application/member/MemberService.java) 변경했어요! 어떻게 생각하시나요?

Reply:
return 코드가 중복되는 느낌이라,,

너잘 의견처럼 변경해도 괜찮을 거 같긴 한데,
웨지 의견처럼 현재 방식이 요구사항 흐름대로 읽히는 것 같아 고민이 되네요...!!

Reply:
이렇게 말하면 너무 강하게 말하는것 같아서 좀 그렇지만, Optional을 사용한 이상 저런식의 코드는 지양하는편이 좋다는 의견을 못 굽히겠어용..
Optional Stream이 한눈에 안읽히는건 익숙하지 않아서 아닐까 하는 생각이 조금 들기도 하는데.. 저는 그냥 보면, 아, 데이터 존재하면 리스폰즈로 만들고 없으면 회원가입 시켜서 리스폰즈 만드는구나 라고 쭉 읽히거든요..

Reply:
```
public MemberLoginResponse loginMember(MemberLoginRequest request) {
        return memberDao.findByName()
                .map(MemberLoginResponse::of)
                .orElseGet(
                        joinMemberAndReturnMemberLoginResponse(request)
                );
    }
```

이런식으로 작성되어 있어도 비즈니스 흐름이 안읽히나요!?

Reply:
```
public MemberLoginResponseDto loginMember(MemberLoginRequestDto requestDto) {
    if (memberDao.exists(requestDto.getName())) {
        return signInMember(requestDto);
    }
    return signUpMember(requestDto);
}

private MemberLoginResponseDto signInMember(MemberLoginRequestDto requestDto) {
    Member member = memberDao.findByName(requestDto.getName())
            .orElseThrow(MemberNotFoundException::new);
    return MemberLoginResponseDto.from(member);
}
```

저는 이 로직에서 `memberDao.exists(requestDto.getName())` 이 메소드 뒤에 `memberDao.findByName(requestDto.getName()).orElseThrow(MemberNotFoundException::new);` 이 로직이 존재한다는 거에서 배드스멜이 먼저 나야한다고 생각합니다.

만일 이런 스멜을 최대한 줄이면서 위 로직을 살리자면
```
public MemberLoginResponseDto loginMember(MemberLoginRequestDto requestDto) {
        Optional<Member> member = memberDao.findByName(requestDto.getName());
        if(member.isPresent()) {
            return MemberLoginResponseDto.from(member.get());
        }
        
        return signUpMember(requestDto);
    }
    
    private MemberLoginResponseDto signUpMember(MemberLoginRequestDto requestDto) {
        Member member = memberDao.insert(requestDto.toMember());
        return MemberLoginResponseDto.from(member);
    }
```
이런 로직이 그나마 괜찮겠죠. 하지만 isPresent() -> get() 은 Optional의 가장 대표적인 안티패턴 중 하나입니다. 따라서 Optional을 사용하면서 이런 절차지향적 로직을 사용하는건 저는 반대하는 입장입니다.

Reply:
@da-nyee @sihyung92 이거 왜 안읽음